{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "4a0d6c38",
   "metadata": {},
   "source": [
    "**Creation of an SQLite Database in python**"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "15858829",
   "metadata": {},
   "source": [
    "*Import necessary library*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "3df051a7-0e4f-4bad-aab3-c4552db0632d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: []"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from sqlite3 import connect\n",
    "connection = connect('worm_genome.sqlite')\n",
    "\n",
    "cursor = connection.cursor()\n",
    "\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "94e835c3",
   "metadata": {},
   "source": [
    "Creation of a feature table "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "id": "ea0b3e19-7317-4ebd-a1f0-a6bcc1fbe127",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done!\n"
     ]
    }
   ],
   "source": [
    "sql = '''\n",
    "CREATE TABLE IF NOT EXISTS features (\n",
    "        feature_id INTEGER PRIMARY KEY AUTOINCREMENT,\n",
    "        seq_id TEXT,\n",
    "        source TEXT,\n",
    "        type TEXT,\n",
    "        start INTEGER,\n",
    "        end INTEGER,\n",
    "        score TEXT,\n",
    "        strand TEXT,\n",
    "        phase TEXT\n",
    "    );\n",
    "'''\n",
    "try:\n",
    "    cursor.execute(sql)\n",
    "except connection.DatabaseError:\n",
    "    print(\"Creating the features table resulted in a database error!\")\n",
    "    connection.rollback()\n",
    "    raise\n",
    "else:\n",
    "    connection.commit()\n",
    "finally:\n",
    "    print(\"done!\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "362e1828",
   "metadata": {},
   "source": [
    "Checking to see what tables i have in my database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "id": "40432fe3-6187-460b-8e54-3caee7d37844",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sqlite_sequence</td>\n",
       "      <td>table</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>attributes</td>\n",
       "      <td>table</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>features</td>\n",
       "      <td>table</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>type_idx</td>\n",
       "      <td>index</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>start_idx</td>\n",
       "      <td>index</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>end_idx</td>\n",
       "      <td>index</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 0      1\n",
       "0  sqlite_sequence  table\n",
       "1       attributes  table\n",
       "2         features  table\n",
       "3         type_idx  index\n",
       "4        start_idx  index\n",
       "5          end_idx  index"
      ]
     },
     "execution_count": 141,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sql = '''\n",
    "SELECT name,type \n",
    "FROM sqlite_master;\n",
    "'''\n",
    "\n",
    "cursor.execute(sql)\n",
    "res_data = cursor.fetchall()\n",
    "res_df = pd.DataFrame(res_data)\n",
    "res_df"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "51b0985a",
   "metadata": {},
   "source": [
    "Creation of an attributes table for my database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "id": "23bec13d-0555-4187-8172-26993dc420f8",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done!\n"
     ]
    }
   ],
   "source": [
    "sql = '''\n",
    "CREATE TABLE IF NOT EXISTS attributes (\n",
    "        attr_id INTEGER PRIMARY KEY AUTOINCREMENT,\n",
    "        feature_id INTEGER,\n",
    "        attr_name TEXT,\n",
    "        value TEXT,\n",
    "        FOREIGN KEY (feature_id) REFERENCES features(feature_id)\n",
    "    );\n",
    "'''\n",
    "try:\n",
    "    cursor.execute(sql)\n",
    "except connection.DatabaseError:\n",
    "    print(\"Creating the features attributes table resulted in a database error!\")\n",
    "    connection.rollback()\n",
    "    raise\n",
    "else:\n",
    "    connection.commit()\n",
    "finally:\n",
    "    print(\"done!\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "497d095c",
   "metadata": {},
   "source": [
    "checks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "id": "5933bfb8-b78d-4ea0-b687-e466a5460364",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>features</td>\n",
       "      <td>table</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sqlite_sequence</td>\n",
       "      <td>table</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>attributes</td>\n",
       "      <td>table</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 0      1\n",
       "0         features  table\n",
       "1  sqlite_sequence  table\n",
       "2       attributes  table"
      ]
     },
     "execution_count": 130,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sql = '''\n",
    "SELECT name,type \n",
    "FROM sqlite_master;\n",
    "'''\n",
    "\n",
    "cursor.execute(sql)\n",
    "res_data = cursor.fetchall()\n",
    "res_df = pd.DataFrame(res_data)\n",
    "res_df"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "ed373727",
   "metadata": {},
   "source": [
    "Creation of indexes for the feature table on the type, start, and end columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "id": "49ffb883-e618-4c53-8e5b-65ad54a38bb0",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sql1 = '''\n",
    "CREATE INDEX IF NOT EXISTS type_idx \n",
    "ON features (type);\n",
    "'''\n",
    "sql2 = '''\n",
    "CREATE INDEX IF NOT EXISTS start_idx \n",
    "ON features (start);\n",
    "'''\n",
    "sql3 = '''\n",
    "CREATE INDEX IF NOT EXISTS end_idx \n",
    "ON features (end);\n",
    "'''\n",
    "cursor.execute(sql1)\n",
    "cursor.execute(sql2)\n",
    "cursor.execute(sql3)\n",
    "connection.commit()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "9262bef5",
   "metadata": {},
   "source": [
    "Creation of indexes for the attributes table on the feature_id and name columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "id": "92248051-1c10-49d0-8c52-300cdedc68d9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sql1 = '''\n",
    "CREATE INDEX IF NOT EXISTS feature_id_idx \n",
    "ON attributes (feature_id);\n",
    "'''\n",
    "sql2 = '''\n",
    "CREATE INDEX IF NOT EXISTS attr_name_idx \n",
    "ON attributes (attr_name);\n",
    "'''\n",
    "cursor.execute(sql1)\n",
    "cursor.execute(sql2)\n",
    "connection.commit()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "65aa7a46",
   "metadata": {},
   "source": [
    "Checks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "id": "19f9a8fb-b7a8-489d-8f83-451788567174",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sqlite_sequence</td>\n",
       "      <td>table</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>attributes</td>\n",
       "      <td>table</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>features</td>\n",
       "      <td>table</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>type_idx</td>\n",
       "      <td>index</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>start_idx</td>\n",
       "      <td>index</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>end_idx</td>\n",
       "      <td>index</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>feature_id_idx</td>\n",
       "      <td>index</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>attr_name_idx</td>\n",
       "      <td>index</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 0      1\n",
       "0  sqlite_sequence  table\n",
       "1       attributes  table\n",
       "2         features  table\n",
       "3         type_idx  index\n",
       "4        start_idx  index\n",
       "5          end_idx  index\n",
       "6   feature_id_idx  index\n",
       "7    attr_name_idx  index"
      ]
     },
     "execution_count": 143,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sql = '''\n",
    "SELECT name,type \n",
    "FROM sqlite_master;\n",
    "'''\n",
    "\n",
    "cursor.execute(sql)\n",
    "res_data = cursor.fetchall()\n",
    "res_df = pd.DataFrame(res_data)\n",
    "res_df"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "b03c57ee-b093-4d41-bf44-f9235d33855c",
   "metadata": {},
   "source": [
    "Creation of a function for parsing the worm genome GFF3 file. \n",
    "    Population of the database after parsing the file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "id": "3456aab2-9995-4fd3-96b7-6ab3a4bb12f6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def parse_gff3_line(line):\n",
    "    parts = line.strip().split('\\t')\n",
    "    if len(parts) < 9:\n",
    "        return None  # Not enough data in this line\n",
    "    seq_id, source, feature_type, start, end, score, strand, phase, attributes = parts[0], parts[1], parts[2], parts[3], parts[4], parts[5], parts[6], parts[7], parts[8]\n",
    "    \n",
    "    # Extracting the name from the attributes\n",
    "    attributes_dict = dict(field.split('=', 1) for field in attributes.split(';') if '=' in field)\n",
    "    \n",
    "    return seq_id, source, feature_type, start, end, score, strand, phase, attributes_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "id": "ce0f3efc-6c59-4eb6-81fb-7a6d47b02471",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "with open('worm_genome.gff3', 'r') as file:\n",
    "    for line in file:\n",
    "        if line.startswith('#'):\n",
    "            continue  # Skip comment lines\n",
    "        parsed_line = parse_gff3_line(line)\n",
    "        if parsed_line is not None:\n",
    "            cursor.execute(\"INSERT INTO features (seq_id, source, type, start, end, score, strand, phase) VALUES (?, ?, ?, ?, ?, ?, ?, ?)\", parsed_line[0:8])\n",
    "            feature_id = cursor.lastrowid  # Get the generated feature_id\n",
    "\n",
    "            # Insert into attributes table\n",
    "            if parsed_line[8] is not None:\n",
    "                attributes_dict = parsed_line[8]\n",
    "                for attr_name, value in attributes_dict.items():\n",
    "                    cursor.execute('''\n",
    "                        INSERT INTO attributes (feature_id, attr_name, value)\n",
    "                        VALUES (?, ?, ?)\n",
    "                    ''', (feature_id, attr_name, value))\n",
    "\n",
    "connection.commit()\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "df67d289",
   "metadata": {},
   "source": [
    "Visualizing the Features table."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "id": "49e3ec36-c555-4be5-b1e2-e14118d1afef",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature_id</th>\n",
       "      <th>seq_id</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>I</td>\n",
       "      <td>WormBase</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1</td>\n",
       "      <td>15072434</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>I</td>\n",
       "      <td>WormBase</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1</td>\n",
       "      <td>15072434</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>I</td>\n",
       "      <td>WormBase</td>\n",
       "      <td>ncRNA_gene</td>\n",
       "      <td>3747</td>\n",
       "      <td>3909</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>I</td>\n",
       "      <td>WormBase</td>\n",
       "      <td>snoRNA</td>\n",
       "      <td>3747</td>\n",
       "      <td>3909</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>I</td>\n",
       "      <td>WormBase</td>\n",
       "      <td>exon</td>\n",
       "      <td>3747</td>\n",
       "      <td>3909</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>659617</th>\n",
       "      <td>659618</td>\n",
       "      <td>X</td>\n",
       "      <td>WormBase</td>\n",
       "      <td>exon</td>\n",
       "      <td>17717328</td>\n",
       "      <td>17717443</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>659618</th>\n",
       "      <td>659619</td>\n",
       "      <td>X</td>\n",
       "      <td>WormBase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>17717328</td>\n",
       "      <td>17717443</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>659619</th>\n",
       "      <td>659620</td>\n",
       "      <td>X</td>\n",
       "      <td>WormBase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>17718435</td>\n",
       "      <td>17718621</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>659620</th>\n",
       "      <td>659621</td>\n",
       "      <td>X</td>\n",
       "      <td>WormBase</td>\n",
       "      <td>exon</td>\n",
       "      <td>17718435</td>\n",
       "      <td>17718720</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>659621</th>\n",
       "      <td>659622</td>\n",
       "      <td>X</td>\n",
       "      <td>WormBase</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>17718622</td>\n",
       "      <td>17718720</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>659622 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        feature_id seq_id    source             type     start       end  \\\n",
       "0                1      I  WormBase       chromosome         1  15072434   \n",
       "1                2      I  WormBase       chromosome         1  15072434   \n",
       "2                3      I  WormBase       ncRNA_gene      3747      3909   \n",
       "3                4      I  WormBase           snoRNA      3747      3909   \n",
       "4                5      I  WormBase             exon      3747      3909   \n",
       "...            ...    ...       ...              ...       ...       ...   \n",
       "659617      659618      X  WormBase             exon  17717328  17717443   \n",
       "659618      659619      X  WormBase              CDS  17717328  17717443   \n",
       "659619      659620      X  WormBase              CDS  17718435  17718621   \n",
       "659620      659621      X  WormBase             exon  17718435  17718720   \n",
       "659621      659622      X  WormBase  three_prime_UTR  17718622  17718720   \n",
       "\n",
       "       score strand phase  \n",
       "0          .      .     .  \n",
       "1          .      .     .  \n",
       "2          .      -     .  \n",
       "3          .      -     .  \n",
       "4          .      -     .  \n",
       "...      ...    ...   ...  \n",
       "659617     .      +     .  \n",
       "659618     .      +     0  \n",
       "659619     .      +     1  \n",
       "659620     .      +     .  \n",
       "659621     .      +     .  \n",
       "\n",
       "[659622 rows x 9 columns]"
      ]
     },
     "execution_count": 151,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sql = '''\n",
    "SELECT * \n",
    "FROM features;\n",
    "'''\n",
    "\n",
    "cursor.execute(sql)\n",
    "res_data = cursor.fetchall()\n",
    "res_df = pd.DataFrame(res_data)\n",
    "header = [row[0] for row in cursor.description]\n",
    "res_df.columns = header\n",
    "res_df\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "188a7813",
   "metadata": {},
   "source": [
    "Visualizing the attributes table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "id": "8a1e7820-3e68-4fc1-95e2-7776f80cfdd8",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>attr_id</th>\n",
       "      <th>feature_id</th>\n",
       "      <th>attr_name</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>ID</td>\n",
       "      <td>chromosome:I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>Alias</td>\n",
       "      <td>BX284601.5,NC_003279.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>ID</td>\n",
       "      <td>gene:WBGene00023193</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>Name</td>\n",
       "      <td>Y74C9A.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>biotype</td>\n",
       "      <td>snoRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3436217</th>\n",
       "      <td>3436218</td>\n",
       "      <td>659621</td>\n",
       "      <td>ensembl_phase</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3436218</th>\n",
       "      <td>3436219</td>\n",
       "      <td>659621</td>\n",
       "      <td>exon_id</td>\n",
       "      <td>cTel55X.1a.e5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3436219</th>\n",
       "      <td>3436220</td>\n",
       "      <td>659621</td>\n",
       "      <td>rank</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3436220</th>\n",
       "      <td>3436221</td>\n",
       "      <td>659621</td>\n",
       "      <td>version</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3436221</th>\n",
       "      <td>3436222</td>\n",
       "      <td>659622</td>\n",
       "      <td>Parent</td>\n",
       "      <td>transcript:cTel55X.1a</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3436222 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         attr_id  feature_id      attr_name                   value\n",
       "0              1           2             ID            chromosome:I\n",
       "1              2           2          Alias  BX284601.5,NC_003279.8\n",
       "2              3           3             ID     gene:WBGene00023193\n",
       "3              4           3           Name                Y74C9A.6\n",
       "4              5           3        biotype                  snoRNA\n",
       "...          ...         ...            ...                     ...\n",
       "3436217  3436218      659621  ensembl_phase                       2\n",
       "3436218  3436219      659621        exon_id           cTel55X.1a.e5\n",
       "3436219  3436220      659621           rank                       5\n",
       "3436220  3436221      659621        version                       1\n",
       "3436221  3436222      659622         Parent   transcript:cTel55X.1a\n",
       "\n",
       "[3436222 rows x 4 columns]"
      ]
     },
     "execution_count": 150,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sql = '''\n",
    "SELECT * \n",
    "FROM attributes;\n",
    "'''\n",
    "\n",
    "cursor.execute(sql)\n",
    "res_data = cursor.fetchall()\n",
    "res_df = pd.DataFrame(res_data)\n",
    "header = [row[0] for row in cursor.description]\n",
    "res_df.columns = header\n",
    "res_df"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "630d07eb-2bb6-4adb-81d9-5f6b36e0c163",
   "metadata": {},
   "source": [
    "SQL Query that retrieves the number of rows in the features table."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "id": "f51927fe-ab81-4a3f-9501-2c4884d465be",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 659622 rows in the table features\n"
     ]
    }
   ],
   "source": [
    "sql = '''\n",
    "SELECT COUNT(*) \n",
    "FROM features;\n",
    "'''\n",
    "cursor.execute(sql)\n",
    "row_count = cursor.fetchall()\n",
    "print(f\"There are {row_count[0][0]} rows in the table features\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "1dc5e812-352a-464b-872c-66cb866e127f",
   "metadata": {},
   "source": [
    "SQL Query that retrieves the number of rows in the attributes table."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "id": "3bfe5882-a375-4b3d-8351-1d015f707b64",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 3436222 rows in the table attributes\n"
     ]
    }
   ],
   "source": [
    "sql = '''\n",
    "SELECT COUNT(*) \n",
    "FROM attributes;\n",
    "'''\n",
    "cursor.execute(sql)\n",
    "row_count = cursor.fetchall()\n",
    "print(f\"There are {row_count[0][0]} rows in the table attributes\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "df6de368-4156-4c38-a595-f8771ef99200",
   "metadata": {},
   "source": [
    "SQL Query that displays the type and the number of features for each type from\n",
    "the features table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "id": "28d3c020-35c2-4698-b801-1d726254277b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>num_features</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CDS</td>\n",
       "      <td>222859</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chromosome</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>exon</td>\n",
       "      <td>269904</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>30960</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>gene</td>\n",
       "      <td>20222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>lnc_RNA</td>\n",
       "      <td>283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>mRNA</td>\n",
       "      <td>33391</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>miRNA</td>\n",
       "      <td>454</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ncRNA</td>\n",
       "      <td>8402</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ncRNA_gene</td>\n",
       "      <td>24765</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>piRNA</td>\n",
       "      <td>15364</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>pre_miRNA</td>\n",
       "      <td>257</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>pseudogene</td>\n",
       "      <td>1791</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>pseudogenic_transcript</td>\n",
       "      <td>1827</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>rRNA</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>snRNA</td>\n",
       "      <td>130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>snoRNA</td>\n",
       "      <td>345</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>tRNA</td>\n",
       "      <td>634</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>28004</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      type  num_features\n",
       "0                      CDS        222859\n",
       "1               chromosome             8\n",
       "2                     exon        269904\n",
       "3           five_prime_UTR         30960\n",
       "4                     gene         20222\n",
       "5                  lnc_RNA           283\n",
       "6                     mRNA         33391\n",
       "7                    miRNA           454\n",
       "8                    ncRNA          8402\n",
       "9               ncRNA_gene         24765\n",
       "10                   piRNA         15364\n",
       "11               pre_miRNA           257\n",
       "12              pseudogene          1791\n",
       "13  pseudogenic_transcript          1827\n",
       "14                    rRNA            22\n",
       "15                   snRNA           130\n",
       "16                  snoRNA           345\n",
       "17                    tRNA           634\n",
       "18         three_prime_UTR         28004"
      ]
     },
     "execution_count": 155,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sql = '''\n",
    "SELECT type, COUNT(*) as num_features\n",
    "FROM features\n",
    "GROUP BY type;\n",
    "'''\n",
    "cursor.execute(sql)\n",
    "res_data = cursor.fetchall()\n",
    "res_df = pd.DataFrame(res_data)\n",
    "header = [row[0] for row in cursor.description]\n",
    "res_df.columns = header\n",
    "res_df"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "4dfbf2e5-58fa-4e1d-9c2a-c2b9430c8aab",
   "metadata": {},
   "source": [
    "SQL Query that displays the number of genes on chromosome 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "id": "5afdb698-7c85-464e-a1d3-051403e36b08",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>COUNT(*)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2889</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   COUNT(*)\n",
       "0      2889"
      ]
     },
     "execution_count": 160,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sql = '''\n",
    "SELECT COUNT(*)\n",
    "FROM features\n",
    "WHERE seq_id = 'I' AND type = 'gene';\n",
    "'''\n",
    "cursor.execute(sql)\n",
    "res_data = cursor.fetchall()\n",
    "res_df = pd.DataFrame(res_data)\n",
    "header = [row[0] for row in cursor.description]\n",
    "res_df.columns = header\n",
    "res_df"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "ec51f30c-196f-4e8d-b67c-9cc5873bcd6f",
   "metadata": {},
   "source": [
    "SQL Query that displays the seq_id and type from the features table and the attr_name\n",
    "and value from the attributes table for all features with the type chromosome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "id": "dd436dc9-6446-4967-a8df-e00eb103f2e1",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_id</th>\n",
       "      <th>type</th>\n",
       "      <th>attr_name</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>I</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>ID</td>\n",
       "      <td>chromosome:I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>I</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>Alias</td>\n",
       "      <td>BX284601.5,NC_003279.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>II</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>ID</td>\n",
       "      <td>chromosome:II</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>II</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>Alias</td>\n",
       "      <td>BX284602.5,NC_003280.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>III</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>ID</td>\n",
       "      <td>chromosome:III</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>III</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>Alias</td>\n",
       "      <td>BX284603.4,NC_003281.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>IV</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>ID</td>\n",
       "      <td>chromosome:IV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>IV</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>Alias</td>\n",
       "      <td>BX284604.4,NC_003282.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>MtDNA</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>ID</td>\n",
       "      <td>chromosome:MtDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>MtDNA</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>Alias</td>\n",
       "      <td>X54252.1,NC_001328.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>V</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>ID</td>\n",
       "      <td>chromosome:V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>V</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>Alias</td>\n",
       "      <td>BX284605.5,NC_003283.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>X</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>ID</td>\n",
       "      <td>chromosome:X</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>X</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>Alias</td>\n",
       "      <td>BX284606.5,NC_003284.9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   seq_id        type attr_name                    value\n",
       "0       I  chromosome        ID             chromosome:I\n",
       "1       I  chromosome     Alias   BX284601.5,NC_003279.8\n",
       "2      II  chromosome        ID            chromosome:II\n",
       "3      II  chromosome     Alias  BX284602.5,NC_003280.10\n",
       "4     III  chromosome        ID           chromosome:III\n",
       "5     III  chromosome     Alias  BX284603.4,NC_003281.10\n",
       "6      IV  chromosome        ID            chromosome:IV\n",
       "7      IV  chromosome     Alias   BX284604.4,NC_003282.8\n",
       "8   MtDNA  chromosome        ID         chromosome:MtDNA\n",
       "9   MtDNA  chromosome     Alias     X54252.1,NC_001328.1\n",
       "10      V  chromosome        ID             chromosome:V\n",
       "11      V  chromosome     Alias  BX284605.5,NC_003283.11\n",
       "12      X  chromosome        ID             chromosome:X\n",
       "13      X  chromosome     Alias   BX284606.5,NC_003284.9"
      ]
     },
     "execution_count": 163,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sql = '''\n",
    "SELECT f.seq_id, f.type, a.attr_name, a.value\n",
    "FROM features f\n",
    "JOIN attributes a ON f.feature_id = a.feature_id\n",
    "WHERE f.type = 'chromosome';\n",
    "'''\n",
    "\n",
    "cursor.execute(sql)\n",
    "res_data = cursor.fetchall()\n",
    "res_df = pd.DataFrame(res_data)\n",
    "header = [row[0] for row in cursor.description]\n",
    "res_df.columns = header\n",
    "res_df"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "0c07541c-069d-473a-a260-beaff6255664",
   "metadata": {},
   "source": [
    "SQL Query that displays the type from the features table and the number of attributes\n",
    "(using also the attributes table) for each feature type that has more than 40 attributes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "id": "f5653a33-ff4a-4b49-858a-97510356ade8",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>num_attributes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CDS</td>\n",
       "      <td>668577</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>exon</td>\n",
       "      <td>2159232</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>30960</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>gene</td>\n",
       "      <td>111530</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>lnc_RNA</td>\n",
       "      <td>1415</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>mRNA</td>\n",
       "      <td>166955</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>miRNA</td>\n",
       "      <td>2270</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ncRNA</td>\n",
       "      <td>42010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ncRNA_gene</td>\n",
       "      <td>123825</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>piRNA</td>\n",
       "      <td>76820</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>pre_miRNA</td>\n",
       "      <td>1285</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>pseudogene</td>\n",
       "      <td>8745</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>pseudogenic_transcript</td>\n",
       "      <td>8925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>rRNA</td>\n",
       "      <td>110</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>snRNA</td>\n",
       "      <td>650</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>snoRNA</td>\n",
       "      <td>1725</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>tRNA</td>\n",
       "      <td>3170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>28004</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      type  num_attributes\n",
       "0                      CDS          668577\n",
       "1                     exon         2159232\n",
       "2           five_prime_UTR           30960\n",
       "3                     gene          111530\n",
       "4                  lnc_RNA            1415\n",
       "5                     mRNA          166955\n",
       "6                    miRNA            2270\n",
       "7                    ncRNA           42010\n",
       "8               ncRNA_gene          123825\n",
       "9                    piRNA           76820\n",
       "10               pre_miRNA            1285\n",
       "11              pseudogene            8745\n",
       "12  pseudogenic_transcript            8925\n",
       "13                    rRNA             110\n",
       "14                   snRNA             650\n",
       "15                  snoRNA            1725\n",
       "16                    tRNA            3170\n",
       "17         three_prime_UTR           28004"
      ]
     },
     "execution_count": 164,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sql = '''\n",
    "SELECT f.type, COUNT(a.attr_id) as num_attributes\n",
    "FROM features f\n",
    "JOIN attributes a ON f.feature_id = a.feature_id\n",
    "GROUP BY f.type\n",
    "HAVING COUNT(a.attr_id) > 40;\n",
    "'''\n",
    "\n",
    "cursor.execute(sql)\n",
    "res_data = cursor.fetchall()\n",
    "res_df = pd.DataFrame(res_data)\n",
    "header = [row[0] for row in cursor.description]\n",
    "res_df.columns = header\n",
    "res_df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
